src/stores/settings.js

```diff
@@ -0,0 +1,76 @@
+import { defineStore } from 'pinia'
+import { fetchApiCall } from '@/utils/api'
+import { useThumbnailsStore } from './thumbnails'
+import { ref } from 'vue'
+
+export const useSettingsStore = defineStore('settings', {
+  state() {
+    return {
+      isColorblindMode: false,
+      selectedImages: [],
+      isConfigLoaded: false,
+      datalabApiBaseUrl: '',
+      datalabArchiveApiUrl: '',
+      observationPortalUrl: '',
+      randomNumbers: [],
+      lineLength: 1,
+      projects: [],
+      largeImageCache: [],
+      smallImageCache: []
+    }
+  },
+  getters: {
+    isSelected: (state) => (image) => {
+      return state.selectedImages.some(selectedImage => selectedImage.basename === image.basename)
+    },
+    // given an image basename, fetches the large res of that image
+    getLargeImageFromBasename: (state) => (imageBasename) => {
+      imageBasename = imageBasename.replace('-small', '')
+      const largeImage = state.largeImageCache.find(obj => obj.basename.replace('-large', '') == imageBasename)
+      if(largeImage){
+        return largeImage
+      }
+      else{
+        console.warn('warn: no large image found with basename:', imageBasename)
+        return null
+      }
+    }
+  },
+  actions: {
+    toggleColorblindMode() {
+      this.isColorblindMode = !this.isColorblindMode
+    },
+    toggleImageSelection(image) {
+      const index = this.selectedImages.findIndex(img => img.basename === image.basename)
+      if (index >= 0) {
+        this.selectedImages.splice(index, 1)
+      } else {
+        this.selectedImages.push(image)
+      }
+    },
+    async loadAndCacheImages(option) {
+      const thumbnailsStore = useThumbnailsStore()
+      const baseUrl = this.datalabArchiveApiUrl + 'frames/'
+      const imageUrl = option ? `${baseUrl}?${option}` : baseUrl
+	
+      const responseData = await fetchApiCall({ url: imageUrl, method: 'GET' })
+      if (responseData && responseData.results) {
+        // Preload all the small thumbnails into the cache. The large thumbnails will be loaded on demand
+        responseData.results.forEach((frame) => {
+          if (option === 'reduction_level=95'){
+            frame.cachedUrl = ref('')
+            thumbnailsStore.cacheImage('small', 'ptr', frame.url, frame.basename).then((cachedUrl) => {
+              frame.cachedUrl.value = cachedUrl
```

Comment:
Maybe I'm misunderstanding but is this .then() callback needed? By this time we've cached the image, right? And we're done with the frame variable?

Reply:
I think `cacheImage()` is returning a new url that points to browser cache. So he's caching the image -> getting a url for the cache -> updating the response object that points to the cache with the right url. 

But I don't understand the need for `frame.cachedUrl = ref('')`
Can you do `frame.cachedUrl = cachedUrl` and delete line 61, making it into a reactive variable?

Reply:
So the v-img components use the `cachedUrl` property as their 'src', so by initially setting it to an empty ref and then updating its value, this triggers the UI to automatically "show" the image once it is available. If you just set the cachedUrl without making it a ref then the image won't show up on initial load if its being cached/created then.